libraries/AP_Math/SCurve.cpp

```diff
@@ -0,0 +1,963 @@
+// calculate the information needed to populate the decreasing jerk magnitude segment from the segment duration tj and jerk magnitude Jm
+// the index variable is the position of this segment in the path and is incremented to reference the next segment in the array
+void SCurve::add_segment_decr_jerk(uint8_t &index, float tj, float Jm)
+{
+    const float Beta = M_PI / tj;
+    const float Alpha = Jm / 2.0f;
+    const float AT = Alpha * tj;
+    const float VT = Alpha * (sq(tj) / 2.0f - 2.0f / sq(Beta));
+    const float PT = Alpha * ((-1.0f / sq(Beta)) * tj + (1.0f / 6.0f) * powf(tj, 3.0f));
+    const float A2T = Jm * tj;
+    const float V2T = Jm * sq(tj);
+    const float P2T = Alpha * ((-1.0f / sq(Beta)) * 2.0f * tj + (4.0f / 3.0f) * powf(tj, 3.0f));
+
+    const float J = Jm;
+    const float T = segment[index - 1].end_time + tj;
+    const float A = (segment[index - 1].end_accel - AT) + A2T;
+    const float V = (segment[index - 1].end_vel - VT) + (segment[index - 1].end_accel - AT) * tj + V2T;
+    const float P = (segment[index - 1].end_pos - PT) + (segment[index - 1].end_vel - VT) * tj + 0.5f * (segment[index - 1].end_accel - AT) * sq(tj) + P2T;
+    add_segment(index, T, SegmentType::NEGATIVE_JERK, J, A, V, P);
+}
+
+// add single S-Curve segment
+// populate the information for the segment specified in the path by the index variable.
+// the index variable is incremented to reference the next segment in the array
+void SCurve::add_segment(uint8_t &index, float end_time, SegmentType seg_type, float jerk_ref, float end_accel, float end_vel, float end_pos)
+{
+    segment[index].end_time = end_time;
+    segment[index].seg_type = seg_type;
+    segment[index].jerk_ref = jerk_ref;
+    segment[index].end_accel = end_accel;
+    segment[index].end_vel = end_vel;
+    segment[index].end_pos = end_pos;
+    index++;
+}
+
+// set speed and acceleration limits for the path
+// origin and destination are offsets from EKF origin
+// speed and acceleration parameters are given in horizontal, up and down.
+void SCurve::set_kinematic_limits(const Vector3f &origin, const Vector3f &destination,
+                                  float speed_xy, float speed_up, float speed_down,
+                                  float accel_xy, float accel_z)
+{
+    // ensure arguments are positive
+    speed_xy = fabsf(speed_xy);
+    speed_up = fabsf(speed_up);
+    speed_down = fabsf(speed_down);
+    accel_xy = fabsf(accel_xy);
+    accel_z = fabsf(accel_z);
+
+    Vector3f direction = destination - origin;
+    const float track_speed_max = kinematic_limit(direction, speed_xy, speed_up, speed_down);
+    const float track_accel_max = kinematic_limit(direction, accel_xy, accel_z, accel_z);
+
+    vel_max = track_speed_max;
+    accel_max = track_accel_max;
+}
+
+// return true if the curve is valid.  Used to identify and protect against code errors
+bool SCurve::valid() const
+{
+    for (uint8_t i = 0; i < num_segs; i++) {
```

Comment:
Should we also assert that `num_segs == max_segs` as we do in some other places?

Reply:
added a check to valid() for the number of segments